### ISSUE TITLE
Provide a way to easely discover kafka service (and route when they will be available) deployed in an Openshift cluster

### ISSUE THREAD

**Issue by valdar** [strimzi/strimzi-kafka-operator#504]
Would be nice to have an easy "discovery mechanism" to get all the Kafka services (and routes when they will be available) deployed in an Openshift cluster.

This would be useful to user friendly configure clients running inside the same Openshift cluster by providing a list of available Kafka broker.

I would have done this by k8s resource filtering but, given the current set of labels applied to strimzi created services I have not found a way to distinguish between them.
Deploying as an example `my-cluster` using the `Kafka ephemeral storage template` results in having services:
- my-cluster-kafka 
- my-cluster-kafka-headless
- my-cluster-zookeeper
- my-cluster-zookeeper-headless

all with the same labels:
```yaml
  labels:
    app: strimzi-ephemeral
    strimzi.io/cluster: my-cluster
    strimzi.io/type: kafka
```
except for  `strimzi.io/name`.

Is quite possible I missing something obvious here and proposing something completely useless but here we go:
a `strimzi.io/subtype` label could be added to discriminate between them; having a different value for each type of service (including Kafka-connect and Kafka-connect-s2i).

**Comment by valdar** [comment]
I had a go to a possible implementation here https://github.com/strimzi/strimzi-kafka-operator/pull/505

**Comment by tombentley** [comment]
The CO itself just uses the name convention to find the various subresources, but I can see that this is less helpful for users than being able to express it with label selectors. I guess a question to ask is whether users need to be able to select any of the subresources, or if they're really just interested the resources they need in order to interact with the cluster (such as the bootstrap service). Do they really need to be able to discover zookeeper resources, for instance?

We should also consider carefully what the names of the services should be (#55 proposes different names).

**Comment by scholzj** [comment]
@valdar I agree with Tom. Could you provide some user stories when you think it would be useful? We actually had something similar on the beginning, but it turned out it brought in more complications to maintain it while not being useful for anything.

The bootstrap service is currently the only think which comes to my mind as being important for discovery. In the future, users will not be allowed to access zookeeper directly. They also cannot do much with any of these resources since any changes they could apply there would conflict with the operator.

**Comment by valdar** [comment]
@scholzj The user story we have to cover in syndesis needs only the Kafka url for connecting clients see: https://github.com/syndesisio/syndesis/issues/2765

I have provided in the PR a different label for each service for completeness, and being future proof.
At the end of the day one label more is not a big deal, but if you want to have that only in the Kakfa service or if you want to restructures services labels all together it's your call.

**Comment by valdar** [comment]
@scholzj any update on this one?

**Comment by scholzj** [comment]
I think we need to first figure out what information should be there for the discovery. I guess we need:
* A label on the bootstrap services which can be used to quesry the service. E.g. `strimzi.io/discovery: true`
* An annotation on the bootstrap services with details about:
  * Ports and their protocol and security settings (TLS). E.g.:

```
strimzi.io/discovery: [ { "port" : 9092, "tls": false, "auth": "none", "protocol": "kafka" }, { "port" : 9093, "tls": true, "auth": "tls", "protocol": "kafka" } ]
```

Would that work? Is it missing some information?

**Comment by valdar** [comment]
I guess this is even more than what we need because actually most of the information present in the annotation is in the service itself isn't it? Like port tls and so on? But I am fine with it.

With your proposed label + annotation what would be the query expressed in words to get all the Strimzi broker in the cluster? "Please k8s give me all the services with the label `strimzi.io/discovery: true`? providing this label will be present only on the bootstrap services?

**Comment by scholzj** [comment]
> I guess this is even more than what we need because actually most of the information present in the annotation is in the service itself isn't it? Like port tls and so on? But I am fine with it.

Well, the port numbers will be there. But the TLS flag would for example need to be decoded from the port names. And the authentication will not be there at all.

> what would be the query expressed in words to get all the Strimzi broker in the cluster?

I do not think you can get that with this label. There are other labels if you want to get all brokers. But why would you need to get such information as list of all brokers? My assumption is that the discovery is interesting for clients. In which case, the bootstrap service is really all you need and what you should use to connect.

**Comment by valdar** [comment]
The use case is on the line of this one: suppose you have an integration platform like Syndesis installed in your cluster along to some instances of Strimzi brokers. When you want to design an integration that reads or writes to kafka, you need to configure your connection. 
You are presented with a form asking for the usual parameters to input, would be helpful for the user if we could populate some of them with what can be discovered from the environment (i.e. k8s.ocp cluster); like populating a dropdown menu' with all the bootstrap services we find (and have access to) in the cluster.
In order to do so, we must be able to query k8s api getting back *all and only* the Strimzi bootstrap services.

**Comment by scholzj** [comment]
Right ... that is the part I understand. But then you asked about how to _"get all the brokers in the cluster when the labels is only on the services"_. Which confused me - the client needs to connect to the bootstrap only and the rest (the individual brokers) will be discovered through the Strimzi protocol. So I do not think you need to be able to find all the brokers, just all the bootstrap services which should be those with the label above.

**Comment by valdar** [comment]
Yes @scholzj you are right, wrong use of words on my part: I ment `bootstrap services`. Sorry about that.

**Comment by scholzj** [comment]
Ok, right ... so in that case yes, you would just search for all services with this label.

**Comment by scholzj** [comment]
@tombentley @ppatierno Do you have any comments to the proposal above?

**Comment by tombentley** [comment]
@scholzj lgtm

**Comment by ppatierno** [comment]
LGTM but just one question. What happens when we have an external listener which creates a new bootstrap service? I mean, should we add this information in the annotation as well?
At this point the drop down menu will show the two discovered bootstrap services and one of them will have "external" in the name but I guess it could be useful for Syndesis to have a better filter on this, to avoid showing external bootstrap services.

**Comment by scholzj** [comment]
I think we should annotate only the internal bootstraps - at least for the time being. The external bootstrap service does not tell you where and how to connect. I think the main purpose of the discovery should by by applications runnign in the same cluster.

**Comment by scholzj** [comment]
Ok, I opened a PR for this to get it finally done. Sorry it took so long.

**Comment by scholzj** [comment]
This should be released in Strimzi 0.16.0.

**Comment by valdar** [comment]
np thanks for this @scholzj.
